mountpoint-s3/src/prefetch/caching_stream.rs

```diff
@@ -132,56 +153,96 @@ where
                 .get_from_client(
                     range.trim_start(block_offset),
                     block_index..block_range.end,
-                    part_queue_producer,
+                    initial_read_window_offset,
+                    part_queue_producer.clone(),
                 )
-                .await;
+                .await
+            {
+                part_queue_producer.push(Err(e));
+            }
+            return;
         }
         // We served the whole range from cache.
         metrics::counter!("prefetch.blocks_served_from_cache").increment(block_range.end - block_range.start);
     }
 
     async fn get_from_client(
-        &self,
+        &mut self,
         range: RequestRange,
         block_range: Range<u64>,
+        initial_read_window_offset: u64,
         part_queue_producer: PartQueueProducer<Client::ClientError>,
-    ) {
-        let key = self.cache_key.key();
+    ) -> Result<(), PrefetchReadError<Client::ClientError>> {
+        let bucket = &self.config.bucket;
+        let cache_key = &self.config.object_id;
         let block_size = self.cache.block_size();
         assert!(block_size > 0);
 
         // Always request a range aligned with block boundaries (or to the end of the object).
         let block_aligned_byte_range =
             (block_range.start * block_size)..(block_range.end * block_size).min(range.object_size() as u64);
+        let request_len = (block_aligned_byte_range.end - block_aligned_byte_range.start) as usize;
+        let block_aligned_byte_range =
+            RequestRange::new(range.object_size(), block_aligned_byte_range.start, request_len);
 
         trace!(
-            ?key,
+            key = self.config.object_id.key(),
             range =? block_aligned_byte_range,
             original_range =? range,
             "fetching data from client"
         );
 
-        let (body_sender, body_receiver) = unbounded();
-        let request_reader_future = read_from_request(
-            body_sender,
-            self.client.clone(),
-            self.bucket.clone(),
-            self.cache_key.clone(),
-            block_aligned_byte_range,
-        );
-        let part_composer_future = compose_parts(
-            body_receiver,
+        let mut part_composer = CachingPartComposer {
             part_queue_producer,
-            self.cache_key.clone(),
-            range,
-            block_range,
-            self.cache.clone(),
-        );
-        join!(request_reader_future, part_composer_future);
+            cache_key: cache_key.clone(),
+            preferred_part_size: self.config.preferred_part_size,
+            original_range: range,
+            block_index: block_range.start,
+            block_offset: block_range.start * block_size,
+            buffer: ChecksummedBytes::default(),
+            cache: self.cache.clone(),
+        };
+
+        // Start by issuing the first request that has a range up to initial read window offset.
```

Comment:
nit: if we had a single function, encapsulating the logic of slitting a request into 2, we won't need to have this duplicated explanation (we could've just use this function in both `ClientPartStream` and `CachingStream`)